packages/smart-contracts/src/contracts/BatchPayments.sol

```diff
@@ -0,0 +1,178 @@
+pragma solidity ^0.8.4;
+
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import "@openzeppelin/contracts/access/Ownable.sol";
+import "./interfaces/ERC20FeeProxy.sol";
+import "./interfaces/EthereumFeeProxy.sol";
+import '@openzeppelin/contracts/security/ReentrancyGuard.sol';
+
+/**
+ * @title BatchPayments
+ * @notice This contract pays multiple paymentrequests in one transaction:
+ *          - on: ERC20, Ethereum
+ *          - regarding ERC20: on multiple tokens in a batch
+ *          - to: multiple addresses
+ */
+contract BatchPayments is Ownable, ReentrancyGuard {
+    
+    // Event to declare a transfer with a reference
+    event TransferWithReferenceAndFee(
+        address tokenAddress,
+        address to,
+        uint256 amount,
+        bytes indexed paymentReference,
+        uint256 feeAmount,
+        address feeAddress
+    );
+    
+    IERC20FeeProxy public paymentErc20FeeProxy;
+    IEthereumFeeProxy public paymentEthereumFeeProxy;
+
+    /**
+     * @param _paymentErc20FeeProxy The address to the ERC20 payment proxy to use.
+     * @param _paymentEthereumFeeProxy The address to the Ethereum payment proxy to use.
+     */
+    constructor(address _paymentErc20FeeProxy, address _paymentEthereumFeeProxy) {
+        paymentErc20FeeProxy  = IERC20FeeProxy(_paymentErc20FeeProxy);
+        paymentEthereumFeeProxy = IEthereumFeeProxy(_paymentEthereumFeeProxy);
+    }
+
+    // Needed because batchEthereumPaymentsWithReferenceAndFee requires that the contract has ethers
+    receive() external payable {}
+
+    /**
+     * @notice Send a batch of erc20 payments, to multiple accounts.
+     * @param _token The token to transact with.
+     * @param _recipients Takes a list of accounts.
+     * @param _amounts Takes a list of amounts, corresponding to recipients[].
+     * @dev Remember to give approval for this contract to spend user tokens.
+     */
+    function batchOrphanERC20Payments(
+        IERC20 _token, 
+        address[] calldata _recipients, 
+        uint256[] calldata _amounts
+    ) external {
+        for (uint256 i = 0; i < _recipients.length; i++)
+            _token.transferFrom(msg.sender, _recipients[i], _amounts[i]);
+    }
+
+    
+    /**
+    * @notice Send a batch of Ethereum payments w/fees with paymentReferences to multiple accounts.
+    *         The sum of _amounts and _feeAMounts must be <= to msg.value.
+    *         If one payment failed, the whole batch is reverted
+    * @param _recipients List of recipients accounts.
+    * @param _amounts List of amounts, corresponding to recipients[].
+    * @param _paymentReferences List of paymentRefs, corr. to the recipients[].
+    * @param _feeAmounts List of amounts of the payment fee, corr. to the recipients[].
+    * @param _feeAddress The fee recipient.
+    * @dev Uses EthereumFeeProxy.sol to pay an invoice and fees, with a payment reference.
+     */
+    function batchEthereumPaymentsWithReferenceAndFee(
```

Comment:
```suggestion
    function batchEthPaymentsWithReferences(
```
New naming convention, "WithFee" is by default.

Reply:
done